File: android-reactive-location/src/main/java/pl/charmas/android/reactivelocation2/observables/MaybeFactory.java
package pl.charmas.android.reactivelocation2.observables;

import io.reactivex.Maybe;
import io.reactivex.MaybeOnSubscribe;
import io.reactivex.MaybeSource;
import io.reactivex.MaybeTransformer;
import io.reactivex.functions.BiPredicate;

public class MaybeFactory {
    private final MaybeContext context;

    public MaybeFactory(MaybeContext context) {
        this.context = context;
    }

    public <T> Maybe<T> createMaybe(MaybeOnSubscribe<T> source) {
        return Maybe.create(source).compose(new RetryOnConnectionSuspension<T>(context.isRetryOnConnectionSuspended()));
    }

    private static class RetryOnConnectionSuspension<T> implements MaybeTransformer<T, T> {
        private final boolean shouldRetry;

        RetryOnConnectionSuspension(boolean shouldRetry) {
            this.shouldRetry = shouldRetry;
        }

        @Override
        public MaybeSource<T> apply(Maybe<T> upstream) {
            if (shouldRetry) {
                return upstream.retry(new IsConnectionSuspendedException());
            }
            return upstream;
        }

        private static class IsConnectionSuspendedException implements BiPredicate<Integer, Throwable> {
            @Override
            public boolean test(Integer integer, Throwable throwable) {
                return throwable instanceof GoogleAPIConnectionSuspendedException;
            }
        }
    }
}

File: android-reactive-location/src/main/java/pl/charmas/android/reactivelocation2/observables/MaybeFactory.kt
package pl.charmas.android.reactivelocation2.observables

import io.reactivex.Maybe
import io.reactivex.MaybeOnSubscribe
import io.reactivex.MaybeSource
import io.reactivex.MaybeTransformer
import io.reactivex.functions.BiPredicate
import pl.charmas.android.reactivelocation2.observables.exceptions.GoogleAPIConnectionSuspendedException

class MaybeFactory(private val context: MaybeContext) {
    fun <T> createMaybe(source: MaybeOnSubscribe<T>?): Maybe<T> {
        return Maybe.create(source).compose(
            RetryOnConnectionSuspension(
                context.isRetryOnConnectionSuspended
            )
        )
    }

    private class RetryOnConnectionSuspension<T> internal constructor(private val shouldRetry: Boolean) :
        MaybeTransformer<T, T> {
        override fun apply(upstream: Maybe<T>): MaybeSource<T> {
            return if (shouldRetry) {
                upstream.retry(IsConnectionSuspendedException())
            } else upstream
        }

        private class IsConnectionSuspendedException :
            BiPredicate<Int, Throwable> {
            override fun test(integer: Int, throwable: Throwable): Boolean {
                return throwable is GoogleAPIConnectionSuspendedException
            }
        }
    }
}